flytekit/annotated/promise.py

```diff
@@ -550,3 +558,128 @@ def __str__(self):
 
     def __repr__(self):
         raise AssertionError(f"Task {self._task_name} returns nothing, NoneType return cannot be used")
+
+
+class NodeOutput(type_models.OutputReference):
+    def __init__(self, node: Node, sdk_type, var: str):
+        """
+        :param node:
+        :param sdk_type: deprecated in mypy flytekit.
+        :param var: The name of the variable this NodeOutput references
+        """
+        self._node = node
+        self._type = sdk_type
+        super(NodeOutput, self).__init__(self._node.id, var)
+
+    @property
+    def node_id(self):
+        """
+        Override the underlying node_id property to refer to SdkNode.
+        :rtype: Text
+        """
+        return self.node.id
+
+    @property
+    def node(self) -> Node:
+        """
+        """
+        return self._node
+
+    @property
+    def sdk_type(self):
+        """
+        """
+        return self._type
+
+    def __repr__(self) -> str:
+        s = f"Node({self.node if self.node.id is not None else None}:{self.var})"
+        return s
+
+
+# TODO we should accept TaskMetadata here and then extract whatever fields we want into NodeMetadata
+def create_and_link_node(
```

Comment:
maybe the name of this function needs to be renamed to reflect that we are Substituting outputs with promises

Reply:
suggestions?